boa_engine/src/string.rs

```diff
@@ -1,474 +1,559 @@
+// - `undocumented_unsafe_blocks` and `missing_safety_doc` requires a `Safety:` section in the
+//   comment or doc of the unsafe block or function, respectively.
+#![deny(
+    unsafe_op_in_unsafe_fn,
+    clippy::undocumented_unsafe_blocks,
+    clippy::missing_safety_doc
+)]
+// Remove when/if https://github.com/rust-lang/rust/issues/95228 stabilizes.
+// Right now this allows us to use the stable polyfill from the `sptr` crate, which uses
+// the same names from the unstable functions of the `std::ptr` module.
+#![allow(unstable_name_collisions)]
+
+use crate::{builtins::string::is_trimmable_whitespace, JsBigInt};
 use boa_gc::{unsafe_empty_trace, Finalize, Trace};
+pub use boa_macros::utf16;
+
 use rustc_hash::{FxHashMap, FxHasher};
 use std::{
-    alloc::{alloc, dealloc, handle_alloc_error, Layout},
+    alloc::{alloc, dealloc, Layout},
     borrow::Borrow,
     cell::Cell,
     hash::BuildHasherDefault,
     hash::{Hash, Hasher},
-    marker::PhantomData,
-    ops::Deref,
-    ptr::{copy_nonoverlapping, NonNull},
-    rc::Rc,
+    ops::{Deref, Index},
+    ptr::{self, NonNull},
+    slice::SliceIndex,
 };
 
-const CONSTANTS_ARRAY: [&str; 419] = [
+/// Utility macro to create a [`JsString`].
+///
+/// # Examples
+///
+/// You can call the macro without arguments to create an empty `JsString`:
+///
+/// ```
+/// use boa_engine::js_string;
+/// use boa_engine::string::utf16;
+///
+/// let empty_str = js_string!();
+/// assert!(empty_str.is_empty());
+/// ```
+///
+///
+/// You can create a `JsString` from a string literal, which completely skips the runtime
+/// conversion from [`&str`] to [`&\[u16\]`]:
+///
+/// ```
+/// # use boa_engine::js_string;
+/// # use boa_engine::string::utf16;
+/// let hw = js_string!("Hello, world!");
+/// assert_eq!(&hw, utf16!("Hello, world!"));
+/// ```
+///
+/// Any `&[u16]` slice is a valid `JsString`, including unpaired surrogates:
+///
+/// ```
+/// # use boa_engine::js_string;
+/// let array = js_string!(&[0xD8AFu16, 0x00A0, 0xD8FF, 0x00F0]);
+/// ```
+///
+/// You can also pass it any number of `&[u16]` as arguments to create a new `JsString` with
+/// the concatenation of every slice:
+///
+/// ```
+/// # use boa_engine::js_string;
+/// # use boa_engine::string::utf16;
+/// const NAME: &[u16]  = utf16!("human! ");
+/// let greeting = js_string!("Hello, ");
+/// let msg = js_string!(&greeting, &NAME, utf16!("Nice to meet you!"));
+///
+/// assert_eq!(&msg, utf16!("Hello, human! Nice to meet you!"));
+/// ```
+#[macro_export]
+macro_rules! js_string {
+    () => {
+        $crate::JsString::default()
+    };
+    ($s:literal) => {
+        $crate::JsString::from($crate::string::utf16!($s))
+    };
+    ($s:expr) => {
+        $crate::JsString::from($s)
+    };
+    ( $x:expr, $y:expr ) => {
+        $crate::JsString::concat($x, $y)
+    };
+    ( $( $s:expr ),+ ) => {
+        $crate::JsString::concat_array(&[ $( $s ),+ ])
+    };
+}
+
+const COMMON_STRINGS: &[&[u16]] = &[
```

Comment:
I would personally move this slice to a submodule, and add a line or two of documentation. It occupies a lot of space at the top of this file.

Reply:
Done!